[flang] Generic type-bound call on a parameterized derived type calls the parent's deferred binding instead of the override when the binding is PRIVATE

### Summary

For a parameterized derived type (PDT) with a `public` generic binding whose specific deferred binding is `private`, flang resolves a call to the generic name to the **parent type's deferred binding** — i.e. to the `abstract interface` symbol — instead of to the overriding binding of the actual type. An abstract interface has no code, so the program fails to link with an undefined reference.

The dispatch tables seem to be generated correctly; only the call site appears to be wrong.

Two independent changes make the problem disappear: declaring the specific binding and its override `public` instead of `private`, or removing the kind parameter from the type (otherwise identical structure).

A private specific binding under a public generic binding is a common and conforming idiom — `private` only prevents calling the specific name directly, it does not prevent generic resolution from selecting it. 

ifx accepts all variants and produces correct results.

### Version

```
$ flang-22 --version
Ubuntu flang version 22.1.2 (1ubuntu1)
Target: x86_64-pc-linux-gnu
Thread model: posix
```

Target: `x86_64-linux-gnu`, linker `ld.bfd`.

### Reproducer

Fails as written. Swapping the two `private` lines for the commented `public` lines makes it work.

```fortran
module m_all
  implicit none
  integer, parameter :: sp = kind(1.0)

  type, abstract :: base_t(rkind)
     integer, kind :: rkind = sp
     integer :: n = 0
   contains
     ! Failing variant:
     procedure(init_sp_i), private, deferred :: init_sp
     ! Working variant:
     !procedure(init_sp_i), public, deferred :: init_sp
     generic, public :: init => init_sp
  end type base_t

  type, extends(base_t) :: impl_t
     real(rkind), allocatable :: mat(:,:)
   contains
     ! Failing variant:
     procedure, private :: init_sp => init_impl_sp
     ! Working variant:
     !procedure, public :: init_sp => init_impl_sp
  end type impl_t

  abstract interface
     subroutine init_sp_i(a, n)
       import :: base_t, sp
       class(base_t(sp)), intent(inout) :: a
       integer, intent(in) :: n
     end subroutine init_sp_i
  end interface

contains

  subroutine init_impl_sp(a, n)
    class(impl_t(sp)), intent(inout) :: a
    integer, intent(in) :: n
    a%n = n
  end subroutine init_impl_sp

end module m_all

program p
  use m_all
  implicit none
  type(impl_t(sp)) :: x
  call x%init(42)
  print *, x%n
end program p
```

**Actual:**

```
$ flang-22 prog.f90 -o prog
/usr/bin/x86_64-linux-gnu-ld.bfd: /tmp/prog-78b2f1.o: in function `_QQmain':
FIRModule:(.text+0x65): undefined reference to `_QMm_allPinit_sp_i'
flang-22: error: linker command failed with exit code 1 (use -v to see invocation)
```

`init_sp_i` is an `abstract interface`; no such symbol can ever be defined.

**Expected:** links and prints `42`, as the `public` variant does.

### FIR analysis

It seems that the dispatch tables are correct and, importantly, **identical** in both variants — the override is properly recorded for `impl_t`:

```mlir
fir.dt_entry "init_sp", @_QMm_allPinit_sp_i deferred    // base_t
fir.dt_entry "init_sp", @_QMm_allPinit_impl_sp          // impl_t
```

Diffing the FIR of the two variants shows that the entire difference is at the call site:

```
$ diff fir_private.mlir fir_public.mlir
380c380
<     %7 = fir.convert %6 : (!fir.box<!fir.type<_QMm_allTimpl_tK4{base_t:!fir.type<_QMm_allTbase_tK4{n:i32}>,mat:!fir.box<!fir.heap<!fir.array<?x?xf32>>>}>>) -> !fir.class<!fir.type<_QMm_allTbase_tK4{n:i32}>>
---
>     %7 = fir.convert %6 : (!fir.box<!fir.type<_QMm_allTimpl_tK4{base_t:!fir.type<_QMm_allTbase_tK4{n:i32}>,mat:!fir.box<!fir.heap<!fir.array<?x?xf32>>>}>>) -> !fir.class<!fir.type<_QMm_allTimpl_tK4{base_t:!fir.type<_QMm_allTbase_tK4{n:i32}>,mat:!fir.box<!fir.heap<!fir.array<?x?xf32>>>}>>
383c383
<     fir.call @_QMm_allPinit_sp_i(%7, %0) fastmath<contract> : (!fir.class<!fir.type<_QMm_allTbase_tK4{n:i32}>>, !fir.ref<i32>) -> ()
---
>     fir.call @_QMm_allPinit_impl_sp(%7, %0) fastmath<contract> : (!fir.class<!fir.type<_QMm_allTimpl_tK4{base_t:!fir.type<_QMm_allTbase_tK4{n:i32}>,mat:!fir.box<!fir.heap<!fir.array<?x?xf32>>>}>>, !fir.ref<i32>) -> ()
409d408
<   func.func private @_QMm_allPinit_sp_i(!fir.class<!fir.type<_QMm_allTbase_tK4{n:i32}>>, !fir.ref<i32>)
```

In the `private` variant, it seems that flang converts the object to the **parent** type `base_tK4` and statically calls the parent's **deferred** entry, emitting an external declaration for a symbol that can never exist. In the `public` variant the object keeps its declared type `impl_tK4` and the override is called directly.

The generic resolution apparently fails to find the `private` specific binding in the scope of the instantiated type and falls back to the inherited deferred binding of the parent.

### Additional data points

* `public` bindings + PDT → works.
* `private` bindings, **no PDT** (same source with the kind parameter removed, `real(sp)`, `class(base_t)`, `type(impl_t)`) → works, and the generated FIR contains no reference to the abstract interface symbol at all.
* `private` bindings + PDT → link failure (this report).
* Only one kind and one specific binding are involved, so no generic disambiguation is required.
* **ifx** compiles, links and produces correct results for every variant above, including `private` + PDT.
* Originally found in a real code base (fypp-generated linear algebra layer with abstract PDT matrix types and per-kind deferred bindings). There, `nm` confirmed that all concrete implementations were present and correctly mangled as `_QM<mod>P<proc>`; only the abstract-interface symbols were referenced and undefined.

Emitting a reference to an abstract interface symbol appears incorrect: flang should resolve the generic to the overriding binding.

--- a/flang/lib/Semantics/expression.cpp
+++ b/flang/lib/Semantics/expression.cpp
@@ -2715,7 +2715,7 @@ auto ExpressionAnalyzer::AnalyzeProcedureComponentRef(
               latest{DEREF(dyType->GetDerivedTypeSpec().typeSymbol().scope())
                          .FindComponent(sym->name())}) {
             if (sym->attrs().test(semantics::Attr::PRIVATE)) {
-              const auto *bindingModule{FindModuleContaining(generic.owner())};
+              const auto *bindingModule{FindModuleContaining(sym->owner())};
               const Symbol *s{latest};
               while (s && FindModuleContaining(s->owner()) != bindingModule) {
                 if (const auto *parent{s->owner().GetDerivedTypeParent()}) {

--- a/flang/test/Semantics/bindings09.f90
+++ b/flang/test/Semantics/bindings09.f90
@@ -1,0 +1,134 @@
+! RUN: %flang_fc1 -fdebug-unparse %s 2>&1 | FileCheck %s
+
+! Ensure that generic resolution for a parameterized derived type uses the
+! module containing the selected binding, not the PDT instantiation site.
+
+module private_pdt
+  implicit none
+  integer, parameter :: sp = kind(1.0)
+
+  type, abstract :: base_t(k)
+    integer, kind :: k = sp
+  contains
+    procedure(private_interface), private, deferred :: binding
+    generic, public :: generic => binding
+  end type
+
+  type, extends(base_t) :: extension_t
+  contains
+    procedure, private :: binding => private_impl
+  end type
+
+  abstract interface
+    subroutine private_interface(x, n)
+      import base_t, sp
+      class(base_t(sp)), intent(inout) :: x
+      integer, intent(in) :: n
+    end subroutine
+  end interface
+
+  type(extension_t(sp)), public :: module_object
+
+contains
+  subroutine private_impl(x, n)
+    class(extension_t(sp)), intent(inout) :: x
+    integer, intent(in) :: n
+  end subroutine
+end module
+
+module third_module
+  use private_pdt
+  implicit none
+contains
+  subroutine call_from_third_module
+    type(extension_t(sp)) :: x
+    ! CHECK: CALL private_impl(x,1_4)
+    call x%generic(1)
+  end subroutine
+end module
+
+module use_renamed
+  use private_pdt, only: renamed_t => extension_t, sp
+  implicit none
+contains
+  subroutine call_use_renamed
+    type(renamed_t(sp)) :: x
+    ! CHECK: CALL private_impl(x,2_4)
+    call x%generic(2)
+  end subroutine
+end module
+
+module private_cross_module
+  use private_pdt
+  implicit none
+
+  ! The inherited private binding cannot be overridden in another module.
+  type, extends(extension_t) :: further_extension_t
+  contains
+    procedure :: binding => unrelated_impl
+  end type
+
+contains
+  subroutine unrelated_impl(x, n)
+    class(further_extension_t(sp)), intent(inout) :: x
+    integer, intent(in) :: n
+  end subroutine
+
+  subroutine call_private_cross_module
+    type(further_extension_t(sp)) :: x
+    ! CHECK: CALL private_impl(x,3_4)
+    call x%generic(3)
+  end subroutine
+end module
+
+module public_pdt
+  implicit none
+  integer, parameter :: sp = kind(1.0)
+
+  type :: base_t(k)
+    integer, kind :: k = sp
+  contains
+    procedure, public :: binding => public_base_impl
+    generic, public :: generic => binding
+  end type
+
+contains
+  subroutine public_base_impl(x, n)
+    class(base_t(sp)), intent(inout) :: x
+    integer, intent(in) :: n
+  end subroutine
+end module
+
+module public_cross_module
+  use public_pdt
+  implicit none
+
+  ! Unlike a private binding, the public binding remains overridable here.
+  type, extends(base_t) :: extension_t
+  contains
+    procedure :: binding => public_extension_impl
+  end type
+
+contains
+  subroutine public_extension_impl(x, n)
+    class(extension_t(sp)), intent(inout) :: x
+    integer, intent(in) :: n
+  end subroutine
+
+  subroutine call_public_cross_module
+    type(extension_t(sp)) :: x
+    ! CHECK: CALL public_extension_impl(x,4_4)
+    call x%generic(4)
+  end subroutine
+end module
+
+program test
+  use private_pdt
+  implicit none
+  type(extension_t(sp)) :: x
+
+  ! CHECK: CALL private_impl(x,5_4)
+  call x%generic(5)
+  ! CHECK: CALL private_impl(module_object,6_4)
+  call module_object%generic(6)
+end program

